### PR TITLE
Fix PAL vertical scaling

### PR DESCRIPTION
**Pull request by iwubcode**
Fixes [10712](https://bugs.dolphin-emu.org/issues/10712).

Vertical scaling is used for PAL games to handle their different aspect ratio.

Previously we were storing the unscaled value in the texture-cache and multiplying the y-scale everytime we needed to compare against the actual height.  Additionally, before creating the texture, we were multiplying the texture's rectangle by the vertical scale.  The rectangle has a subtle +1 added to its height and this was causing the vertically scaled value to overflow which was causing a crash ingames were actually expecting in memory.

This PR scales before applying the "+1" and then stores that value into the texture cache.  This fixes the crashes and removes a number of unnecessary multiplications we were doing!

### CODE REVIEW

File: Source/Core/VideoCommon/BPStructs.cpp [hunk 1]
```diff
@@ -252,12 +252,14 @@ static void BPWritten(const BPCmd& bp)
 
       float num_xfb_lines = 1.0f + bpmem.copyTexSrcWH.y * yScale;
 
+      srcRect.bottom = (int)(bpmem.copyTexSrcXY.y + num_xfb_lines);
```
Comment (BhaaLseN) [inline Source/Core/VideoCommon/BPStructs.cpp]: `static_cast<int>` please instead of a C-style cast.

**Comment by JMC47** [comment]
Did you see the commit I linked in HybridXFB detailing the same crash happening in the past?

**Comment by iwubcode** [comment]
@JMC47 - yes I saw it.  That code line did not change in Hybrid XFB.  However, I took the concept of instead of `yScale * (height + 1)` to instead do `(yScale * height) + 1` in another location and this PR is the result.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 2]
```diff
@@ -1744,7 +1741,13 @@ void TextureCacheBase::CopyRenderTargetToTexture(u32 dstAddr, EFBCopyFormat dstF
       entry->may_have_overlapping_textures = false;
       entry->is_custom_tex = false;
 
-      CopyEFBToCacheEntry(entry, is_depth_copy, srcRect, scaleByHalf, dstFormat, isIntensity);
+      // For XFB, the resulting XFB copy texture is the height of the actual texture + the y-scaling
+      // however, the actual Wii/GC texture that we want to "stretch" to this copy is without the y-scaling
+      // so we need to remove it here
+      // The best use-case for this is PAL games which have a different aspect ratio
+      EFBRectangle src_rect = srcRect;
```
Comment (lioncash) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: This should *probably* be given a name different from the one `srcRect` should actually be ;)

**Comment by JosJuice** [comment]
With this PR, Baten Kaitos PAL seems to be working again.

**Comment by JMC47** [comment]
All of my PAL games are booting again.  Good stuff!

**Comment by JMC47** [comment]
LGTM

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 3]
```diff
@@ -1744,7 +1739,13 @@ void TextureCacheBase::CopyRenderTargetToTexture(u32 dstAddr, EFBCopyFormat dstF
       entry->may_have_overlapping_textures = false;
       entry->is_custom_tex = false;
 
-      CopyEFBToCacheEntry(entry, is_depth_copy, srcRect, scaleByHalf, dstFormat, isIntensity);
+      // For XFB, the resulting XFB copy texture is the height of the actual texture + the y-scaling
+      // however, the actual Wii/GC texture that we want to "stretch" to this copy is without the
+      // y-scaling so we need to remove it here
+      // The best use-case for this is PAL games which have a different aspect ratio
+      EFBRectangle unscaled_rect = srcRect;
+      unscaled_rect.bottom /= y_scale;
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: May this introduce a round trip error? I'd suggest to pass width and scaled height as new arguments and to not modify the source_rect.

**Comment by degasus** [comment]
Let me merge this now as it is, but I think we should swap the behavior of source_rect and native_height again.

**Comment by degasus** [comment]
@iwubcode May you review #6268 ?

**Comment by 2kDD** [comment]
After the changes are incorrectly displayed, the Resident Evil 4 Wii [PAL] in 16:9 aspect ratio. At 4:3 fine.

**Comment by TheAxeC** [comment]
I'm on MacOS High Sierra Version 10.13.5 on a MacBook Po (Retina, 15-inch, Mid 2015) using Dolphin 5.0-6084 (which refers to this Pull request). When running any EU game (for example, Tom Clancy's Rainbow Six Lockdown or Gladius or Resident Evil 4) there are bright blue bars (top of the screen, vertical, fully colored blue). Previous versions of Dolphin (such as Dolphin 5.0-6081) do not have this issue. This issue persists until the latest builds.

**Pull request merged**